Use `cosmiconfig` Library for Contextual Config File Loading

bendvc opened the pull request:
# Description

During development of multi-site support in the PWA starter project we've identified a productizable feature that can added to the SDK, in this case it's targeted loading of configuration files. By using the `cosmiconfig` library to load configuration files we can customize what file is used and when. 

For example: If you want to have one configuration file for your production environment and one for your development environment, you can do that.


# Changes

1. Added `cosmiconfig` package to the SDK and used it in the set up of the express application. Previously used the package.json to set the `mobify` option.
2. Surface the configuration to the React App via the react rendering pipeline, serializing the config in the html as the `__CONFIG__` global.
3. Defined `cosmiconfig` "searchPlaces" to prioritize configuration loading from the deployment target (environment) named file with fallbacks from `local`, and default definitions to definitions in the package.json file.
4. Allow for config files to be writing in yml so we can support comments.
5. Added json endpoint to get/validate config for environments.

# Future Work

1. Add `hooks` and `providers` to the SDK for getting and using the config object.
2. Move other commerce and einstein configs in to the new default.json config file.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

# 3pp Approval

- [cosmiconfig](https://gus.lightning.force.com/lightning/r/ADM_Third_Party_Software__c/a0qEE0000001uRZYAY/view)
- 
> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

bendvc committed:
Super simple config implementation.

bendvc committed:
Clean up implementation

bendvc committed:
Use a different snyc config package

bendvc committed:
Add notes, remove dead code.

bendvc committed:
Make it backwards compatible

bendvc committed:
example customization of moduleName resolution

bendvc committed:
Added some notes

bendvc committed:
More notes

bendvc committed:
Move work into SDK

bendvc committed:
Merge branch 'develop' into cosmiconfig-example

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -11,6 +11,8 @@ import {proxyConfigs} from '../../utils/ssr-shared'
 
 const onClient = typeof window !== 'undefined'
 
+let _config
```
Do not do this, this will leak on configuration for all clients making requests to the lambda. Normally this would be desirable because there is a 1-1 mapping of a lambda to a configuration. But that only holds true when using the default logic for loading your config, it breaks if you were to customize the config file resolving to base your choice from the url path for example.

bendvc commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +53,59 @@ export const getProxyConfigs = () => {
+ */
+export const getConfig = (opts = {}) => {
+    const {fileNameResolver} = opts
+    if (_config) {
```
Bad. See above.

alexvuong commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +53,59 @@ export const getProxyConfigs = () => {
+            `config/${moduleName}.yaml`,
+            `config/${moduleName}.json`,
+            `config/local.yml`,
+            `config/local.yaml`,
```
The moduleName will always be `default` when there is no resolver or `DEPLOY_TARGET` is undefined. In this case, the local.json will always be ignored. We probably don't really need `DEFAULT_CONFIG_MODULE_NAME` imo

bendvc replied:
Good catch.. resolved it.

bendvc committed:
Move configuration implementation into the express server layer

bendvc committed:
Clean up config files

bendvc committed:
Semi-fix upload script, remove customizability of config file resolving.

bendvc committed:
Remove dubplicate `getConfig` implementation.

bendvc committed:
Fix lint errors

bendvc committed:
Make tests pass

bendvc committed:
Use regex for pink and config endpoints..

to make things consistent looking. Remove default value for config module name as it broke the resolving logic expected. Mock full defualt config to window for testing.

bendvc committed:
Commenting

bendvc committed:
Fix bug with prepare-dist script

bendvc committed:
Merge branch 'develop' into cosmiconfig-example

bendvc committed:
Rebuild lock files, clean up config files

bendvc commented on packages/pwa-kit-react-sdk/bin/upload.js:
```diff
@@ -62,10 +62,12 @@ try {
     Utils.fail(message)
 }
 
+const config = Utils.getConfig(argv.t)
```
`getConfig` currently doesn't accept any arguments. It looks like we are attempting to get the config object for a given target, that will need to be fixed.

bendvc commented on packages/pwa-kit-react-sdk/scripts/utils.js:
```diff
@@ -243,4 +245,6 @@ Utils.requestErrorMessage = {
     code500: 'Internal Server Error. Please report this to Salesforce support team.'
 }
 
+Utils.getConfig = getConfig
```
This is simply a re-export of the `getConfig` util so they are all accessible via this one file when used in the binaries.

bendvc committed:
Rename __APPCONFIG__ to __APP_CONFIG__

bendvc committed:
Clean up

bendvc committed:
Variable name changes

bendvc committed:
Remove example api config entries

bendvc committed:
Insure that `config` files are added to the bundle

respect the .gitignore

bendvc committed:
Remove testing config entires

bendvc committed:
Use correct config in pwa-build script

bendvc committed:
Fix cosmiconfig deps

bendvc committed:
Fix config search for remote

bendvc committed:
Fix tests, add changelog update

bendvc committed:
Merge branch 'develop' into cosmiconfig-example

bendvc committed:
Update package-lock.json

bendvc committed:
Fix typo

kevinxh commented on packages/pwa-kit-react-sdk/package.json:
```diff
@@ -156,6 +157,7 @@
     "@loadable/component": "^5.15.0",
     "@loadable/server": "^5.15.0",
     "@loadable/webpack-plugin": "^5.15.0",
+    "cosmiconfig": "^7.0.1",
```
curious why is this a peerdependency?

olibrook replied:
I kinda agree with Kevin – if the PWA doesn't import `cosmiconfig` directly (doesn't look like it to me), then I'd expect `cosmicconfig` to be a `dependency`. Calling code in the `pwa` would then only touch our `getConfig` function...

bendvc replied:
Totally agree with both of you. Surprisingly doing this solved a problem I was having. I was waiting on review to get confirmation that this was not what I wanted. So I'll go back and give it another attempt as getting it working as a normal dependency.

bendvc replied:
The problem lies in the `pwa-build` script. Specifically when it determines which node modules need to be installed for the bundle. It's only ever referencing the package.json in the `pwa` for version numbers of the externals and not the package.json of the `sdk`. 

Thats why having `express` as an external works, because it's a peer dependency of the sdk and shows up in the dependency list of the pwa. But `cosmiconfig` doesn't get installed if it wasn't a peer dependency. 

Lets have a quick convo about the implications of changing the script to look at the `sdk` package.json file as well.

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/server/express.js:
```diff
@@ -212,6 +212,13 @@ export const createApp = (options) => {
     // Attach built in routes and middleware
 
+    // Attach the application configuration to the request object.
+    app.use((req, _, next) => {
+        req.config = options.mobify
+
+        next()
+    })
```
here we are attaching configs at the request level, by using a middleware. I'm wondering could we do it at the express app level like:

```suggestion
    app.config = options.mobify
```

Personally I think this is easier to reason about since the configurations are bound to an environment. My assumption here is that you will always have the same configuration regardless of the request.

kevinxh replied:
We should only load the config once, at server boot up time.

bendvc replied:
Totally agree. I'll make this change

olibrook commented on packages/pwa-kit-react-sdk/bin/pwa-build.js:
```diff
@@ -19,16 +20,17 @@ const webpack = path.join('node_modules', '.bin', 'webpack')
 sh.set('-e')
 sh.config.silent = true
 
+const config = Utils.getConfig() || {}
```
I don't have time to review this thoroughly this minute, but these changes will affect V2. Can we sync up on the changes before merging please @bendvc ?

bendvc replied:
Small summary of what we talked about IRL. Oli's concern was whether or not we can implement in v2 and if it's possible to maintain the same API for getting configuration in V2 as we have in this PR. This resulted in my doing a small spike where I was able to get configuration integrated quite easily, the API however is a little different based on how the new mono repo is broken into more projects, the express server definition doesn't exist along side of the react rendering and overall having the utility split up felt more natural to me. It's probably possible to get the same API using webpack plugins to include the cosmiconfig module, but that isn't ideal.

olibrook commented on packages/pwa-kit-react-sdk/src/ssr/server/express.js:
```diff
@@ -258,13 +265,21 @@ export const createApp = (options) => {
+            .set('cache-control', NO_CACHE)
+            .json(req.config)
+            .end()
+    )
```
If this endpoint isn't used by the code in this PR (I can't see that it is anywhere), then I would recommend not exposing it for two reasons:

1. It's extra stuff that were expected to support on an ongoing basis, without a use-case.
2. I believe people would be surprised to see that the content of a config file is immediately made available on a public API endpoint.

I know that it would get embedded in public code anyway, even without this endpoint, but do you think that exposing configs like this would be surprising from a security perspective?

bendvc replied:
This was used for testing. I've since removed it.

olibrook commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+    // Synchronously load the `cosmiconfig` so we don't get errors from webpack in the
+    // browser bundle.
+    const _require = eval('require')
+    const {cosmiconfigSync} = _require('cosmiconfig')
```
@bendvc We need to sync about with this bit about V2, too! There are changes on the way that I think will be problematic to resolve here.

bendvc replied:
As discussed in one of the above [comments](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/390#discussion_r813168156) there was a spike for V2. As pointed out by Oli these few lines are the crux of what would be difficult in V2 since the notion of externals is no longer there. So we would have to think oh how to include the 'cosmiconfig' module without having it compiled for the client. These did it with these 3 lines and externals, but the same solution doesn't work in v2.

olibrook commented on packages/pwa/app/ssr.js:
```diff
@@ -29,8 +32,9 @@ const app = createApp({
-    mobify: WEBPACK_PACKAGE_JSON_MOBIFY,
+    // provided by a webpack DefinePlugin. Alternatively the object can be defined
+    // in a configuration file in the `config` folder.
+    mobify: config,
```
Tasty.

kevinxh commented on packages/pwa/config/default.json:
```diff
@@ -0,0 +1,39 @@
+{
```
I'm a bit confused by the name of this file, I thought the name of the config is the mrt environment id or `local`. When is this default get used?

kevinxh replied:
after reading thru the PR i think i get it now.

kevinxh commented on packages/pwa-kit-react-sdk/src/webpack/config.js:
```diff
@@ -187,7 +187,16 @@ const common = {
+            patterns: [
+                {from: 'app/static/', to: 'static/'},
+                {
+                    from: 'config/',
+                    to: 'config/',
```
nit, shall we follow the existing folder naming pattern (using plurals like `/components`, `/pages`) and rename the folder to `/configs`?

bendvc replied:
I think config "singular" is an established pattern. Check your home folder for a `.config` folder. There is probably one there and it hold configurations for all sorts of things.

kevinxh commented on packages/pwa-kit-react-sdk/bin/upload.js:
```diff
@@ -62,10 +62,19 @@ try {
+// Get the config object for the current target, if no target was set
+// it will return the defualt configuration and that one will be
+// validated.
+const config = Utils.getConfig()
```
Shall we remove the code in this file where it tries to read package.json and grabs value from there?

I'm talking about the code [here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/pwa-kit-react-sdk/bin/upload.js#L49-L63).

bendvc replied:
The `package.json` file is still being used for project slug is one isn't provided via the command line.

kevinxh commented on packages/pwa-kit-react-sdk/bin/upload.js:
```diff
@@ -62,10 +62,19 @@ try {
+process.env.DEPLOY_TARGET = argv.t
+
+// Get the config object for the current target, if no target was set
+// it will return the defualt configuration and that one will be
```
```suggestion
// it will return the default configuration and that one will be
```

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+    const {config} = explorerSync.search() || {}
+
+    if (!config) {
+        throw new Error('Application configuration not found!')
```
nit, could we improve the error message to include the specific file locations like:

```js
const searchPlaces = [
            `config/${moduleName}.yml`,
            `config/${moduleName}.yaml`,
            `config/${moduleName}.json`,
            `config/local.yml`,
            `config/local.yaml`,
            `config/local.json`,
            `config/default.yml`,
            `config/default.yaml`,
            `config/default.json`,
            `package.json`
        ]
throw new Error(`Application configuration not found!

Possible configuration file locations:

${searchPlaces.join('\n')}
`)
```

Preview of the error:
![Screen Shot 2022-02-22 at 1 24 16 PM](https://user-images.githubusercontent.com/10948652/155221927-adbe38dd-778c-4c89-a7ca-a4aff11af2f3.png)

bendvc replied:
Good idea. I've added this.

kevinxh commented on packages/pwa/app/ssr.js:
```diff
@@ -29,8 +32,9 @@ const app = createApp({
 
     // This is the value of the 'mobify' object from package.json
-    // provided by a webpack DefinePlugin
-    mobify: WEBPACK_PACKAGE_JSON_MOBIFY,
+    // provided by a webpack DefinePlugin. Alternatively the object can be defined
+    // in a configuration file in the `config` folder.
```
i think the first half of this comment is no longer applicable. We no longer use the `DefinePlugin`, am i right?

bendvc committed:
Remove config and ping endpoints

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+    }
+
+    const isRemote = Object.prototype.hasOwnProperty.call(process.env, 'AWS_LAMBDA_FUNCTION_NAME')
+    let moduleName = process?.env?.DEPLOY_TARGET || ''
```
does it make sense to add an argument to `getConfig` for moduleName and fallback to `process?.env?.DEPLOY_TARGET`? since this util function could run in a local dev process where `process?.env?.DEPLOY_TARGET` is not defined, such as the bundle upload script.

bendvc replied:
I think having an argument would make the function less universal in a way. It would be an argument that only makes sense on the server. That is why I shy'd away from it. With regard to the upload script I do set the environment variable based on the target argument if one exists.

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+    if (onClient) {
+        return (
+            window.__CONFIG__ ||
+            JSON.parse(document.getElementById('mobify-data').innerHTML).__CONFIG__
```
We briefly talked about not running this expensive operation `JSON.parse`. The plan is to remove the `JSON.parse` call

bendvc replied:
I'll add a comment explaing what is happening here a little better.

kevinxh commented on packages/pwa-kit-react-sdk/bin/upload.js:
```diff
@@ -62,10 +62,19 @@ try {
     Utils.fail(message)
 }
 
+// Set the deployment target env var, this is required to ensure we
+// get the correct configuration oject.
+process.env.DEPLOY_TARGET = argv.t
+
+// Get the config object for the current target, if no target was set
+// it will return the defualt configuration and that one will be
+// validated.
+const config = Utils.getConfig()
```
(related to https://github.com/SalesforceCommerceCloud/pwa-kit/pull/390#discussion_r812373437)

instead of passing the target implicitly thru environment variable, would it be better if we add an optional argument to `getConfig`, something like:
```suggestion

// Get the config object for the current target, if no target was set
// it will return the defualt configuration and that one will be
// validated.
const config = Utils.getConfig(argv.t)
```

bendvc replied:
I kind of touched on this above a little. I'm a little on the fence. If we do something like passing in a argument, it means that the description of the function argument is somwthing like 'moduleName: is the file name of the configuration you want to load, this argument isn't used on the client and will be ignored'. 

Another argument for keeping things the way they are, is that if we pass in an argument for the file to load, it somewhat devalues to structure of the 'searchPlaces` option, a user can then use this method to load any config file they want. When we are really setting up rules to configuration saying, I'll load the config for your environment, if thats not found I'll load x, y, then z. 

If we are to change the function to accept a file name, it is more like this function loads a json file with the name you provide from the config folder, if it doesn't exist we'll load x, y, then z. 

I like dependency injection, but I think in this case we want the simple "loads the config for your environment" description.

bendvc committed:
Attach config to app, not via middleware

bendvc committed:
Remove test for ping endpoint

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+    // Match config files based on the specificity from most to most general.
+    const explorerSync = cosmiconfigSync(moduleName, {
+        packageProp: 'mobify',
+        searchPlaces: [
```
I don't see `.js` extension here, do we want to support commonjs modules? I'm under the impression that cosmiconfig supports commonjs modules.

bendvc replied:
It does. At the moment tho I don't think this is in the acceptance criteria so I left that complication out. We can easily add it in later.

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,64 @@ export const getProxyConfigs = () => {
+ * @returns - the application configuration object.
+ */
+/* istanbul ignore next */
+export const getConfig = () => {
```
I'm a little bit concerned about the performance of this util function, specifically on the server side.

This util could potentially be called from user land many times during ssr, and each time this util gets called, it will dynamically require cosmiconfig and do a `cosmiconfigSync` file search. I'm not exactly sure the speed magnitude of the operations, but I'd imagine there is a risk that it could cause a somewhat significant slowdown.

is it possible to cache it, so that we technically only load configs **once** at server bootup.

bendvc replied:
`cosmiconfig` has built in caching. You can read about it [here](https://www.npmjs.com/package/cosmiconfig#caching). I think it does what we need it to do.

bendvc committed:
Update packages/pwa-kit-react-sdk/bin/upload.js

Co-authored-by: Kevin He <kevin.he@salesforce.com>

bendvc committed:
Print config search locations on error

bendvc committed:
Merge branch 'cosmiconfig-example' of https://github.com/SalesforceCommerceCloud/pwa-kit into cosmiconfig-example

bendvc committed:
Added note about config parsing.

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.js:
```diff
@@ -51,3 +51,73 @@ export const getProxyConfigs = () => {
+
+    // Match config files based on the specificity from most to most general.
+    const explorerSync = cosmiconfigSync(moduleName, {
+        packageProp: 'mobify2',
```
```suggestion
        packageProp: 'mobify',
```

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/universal/utils.server.test.js:
```diff
@@ -7,7 +7,7 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
-
+/* eslint-disable header/header */
```
this doesn't seem right. i don't see any other changes in this file. is this intended?

bendvc commented:
Closing in favour of this [PR](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/421)
